R/MsBackendTimsTof-functions.R

```diff
@@ -138,16 +144,50 @@ MsBackendTimsTof <- function() {
         if (!length(sd <- setdiff(columns, c("frame", "scan"))))
             stop("At least one column value different from 'frame' and",
                  " 'scan' required")
-        tmp <- query(tms, unique(x@indices[I, "frame"]), c("frame", "scan", sd))
-        f <- factor(paste(tmp$frame, tmp$scan),
-                    levels = paste(x@indices[I, "frame"], x@indices[I, "scan"]))
-        if (length(sd) == 1)
-            res[I] <- unname(split(tmp[, sd], f))
-        else res[I] <- unname(split.data.frame(as.matrix(tmp[, sd]), f))
+        ## Ensure mz and intensity are always first, if provided
+        sd <- c(intersect(c("mz", "intensity"), sd),
+                setdiff(sd, c("mz", "intensity")))
+        i_frame <- x@indices[I, "frame"]
+        tmp <- query(tms, unique(i_frame), c("frame", "scan", sd))
+        ## subset tmp if we're about to extract only few scans.
+        if (length(i_frame) < (nrow(tmp) / 10)) {
```

Comment:
This `if` statement subsets `tmp` if only few scans are to be extracted. This can result in considerably faster `split.data.frame` and `paste` calls below.

Reply:
the `query` function allows only to retrieve data filtering by frame but not by spectrum. Thus, even if we want to extract a single spectrum we need to query all spectra in a frame and then we have that potentially large `data.frame` we need to split. I realized that it can be much faster if in such cases we first subset the `data.frame` to scan IDs we're interested in and then split it.

So, if we extract at least 10 times less scans then we have data, we first subset. Could also be that this is an overoptimization - or even wrongly implemented?

Reply:
I'm sorry I got confused erroneously thinking `length(i_frame)` as the number of frames but it is actually the number of scans. So basically we are subsetting when it is advantageous. I think it makes sense! Thank you!

Reply:
yes, I know - it's confusing :)